weave/trace/annotation_parser.py

```diff
@@ -0,0 +1,151 @@
+
+    Returns:
+        ContentFileAnnotation | None: An instance of ContentFileAnnotation if the parsing is successful,
+    """
+    # Content annotation without a format specifier
+    # Example: "typing.Annotated[SomePathLikeType, <class 'weave.type_handlers.Content.content.Content'>]"
+    # Explanation:
+    # - r"typing\.Annotated\[": Matches the literal "typing.Annotated["
+    # - r"(.+?)": Group 1. Captures the base type (non-greedy).
+    # - r",\s*": Matches a comma followed by optional whitespace.
+    # - fr"<class '{re.escape(content_class_name)}'>": Matches the class representation.
+    # - r"\]": Matches the closing square bracket for Annotated.
+    pattern_without_format = re.compile(
+        r"typing\.Annotated\["
+        r"(.+?),\s*"  # Group 1: Base type (non-greedy)
+        rf"<class '{re.escape(CONTENT_CLASS_NAME)}'>"
+        r"\]"  # Closing bracket for Annotated[...]
+    )
+
+    match_without_format = pattern_without_format.fullmatch(annotation_string)
+
+    if not match_without_format:
+        return None
+
+    return ContentAnnotation(
+        base_type=match_without_format.group(1).strip(),
+        media_class=CONTENT_CLASS_NAME,
+        raw_annotation=annotation_string,
+    )
+
+
+def parse_content_annotation(
+    annotation_string: str,
+) -> ContentAnnotation | None:
+    """
+    Parses an content type annotation string.
+
+    The function expects the string to be one of two forms:
+    1. typing.Annotated[SomeType, weave.type_wrappers.Content.content.Content[typing.Literal['format']]]
+    2. typing.Annotated[SomeType, <class 'weave.type_wrappers.Content.content.Content'>]
+
+    It extracts the base type (which can be complex, like typing.Union),
+    confirms the Content class, and identifies the content format.
+
+    Args:
+        annotation_string: The string representation of the type annotation.
+
+    Returns:
+        A dictionary containing:
+        - "base_type": The underlying type being annotated (e.g., "SomeType", "str", "typing.Union[str, bytes]").
+        - "content_class": The name of the content handler class.
+        - "content_format": The content extension or mimetype (e.g. "mp3", "audio/mpeg") if specified, otherwise None.
+        - "raw_annotation": The original input string.
+        If parsing fails, it returns a dictionary with an "error" key and "raw_annotation".
+    """
+    # Try matching the pattern with format first (it's more specific)
+    return parse_data_annotation(annotation_string) or parse_file_annotation(
+        annotation_string
+    )
+
+
+def parse_from_signature(sig: Signature) -> dict[str, ContentAnnotation]:
+    parsed_annotations = {}
+
+    for param_name, param in sig.parameters.items():
+        if not param.annotation:
+            continue
+
+        parse_result = parse_content_annotation(str(param.annotation))
+
+        if not parse_result:
+            continue
```

Comment:
nit: could be walrus

Reply:
Fixed. I'm going to leave the continue above since it looks cleaner than the other ways of checking.